core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -73,115 +69,66 @@
  * @since 11.0
  */
 @Scope(Scopes.NAMED_CACHE)
-public class DefaultIracManager implements IracManager, Runnable {
+public class DefaultIracManager implements IracManager {
 
    private static final Log log = LogFactory.getLog(DefaultIracManager.class);
    private static final boolean trace = log.isTraceEnabled();
 
    @Inject RpcManager rpcManager;
-   @Inject Configuration config;
    @Inject TakeOfflineManager takeOfflineManager;
    @Inject ClusteringDependentLogic clusteringDependentLogic;
    @Inject CommandsFactory commandsFactory;
    @Inject IracVersionGenerator iracVersionGenerator;
 
-   private final Map<Object, Object> updatedKeys;
-   private final Semaphore senderNotifier;
-   private volatile ExponentialBackOff backOff;
+   private final Map<Object, State> updatedKeys;
+   private final Collection<XSiteBackup> asyncBackups;
+   private final IracExecutor iracExecutor;
    private volatile boolean hasClear;
-   private volatile Collection<XSiteBackup> asyncBackups;
-   private volatile Thread sender;
-   private volatile boolean running;
 
-   public DefaultIracManager() {
+   public DefaultIracManager(Configuration config) {
       this.updatedKeys = new ConcurrentHashMap<>();
-      this.senderNotifier = new Semaphore(0);
-      this.backOff = new ExponentialBackOffImpl();
+      this.iracExecutor = new IracExecutor(this::run);
+      this.asyncBackups = asyncBackups(config);
    }
 
-   private static Collection<XSiteBackup> asyncBackups(Configuration config, String localSiteName) {
+   private static Collection<XSiteBackup> asyncBackups(Configuration config) {
       return config.sites().asyncBackupsStream()
-            .filter(bc -> !localSiteName.equals(bc.site()))
             .map(bc -> new XSiteBackup(bc.site(), true, bc.replicationTimeout())) //convert to sync
             .collect(Collectors.toList());
    }
 
-   private static Stream<?> keyStream(WriteCommand command) {
-      return command.getAffectedKeys().stream();
-   }
-
-   private static boolean backupToRemoteSite(WriteCommand command) {
-      return !command.hasAnyFlag(FlagBitSets.SKIP_XSITE_BACKUP);
-   }
-
    private static IntSet newIntSet(Address ignored) {
       return IntSets.mutableEmptySet();
    }
 
+   @Inject
+   public void inject(@ComponentName(KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService executorService,
+                      @ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor) {
```

Comment:
Can't this be the non blocking executor? Or does something block in `DefaultIracManager#run`?

Reply:
It peeks the DataContainer & Persistence (if enabled). I didn't want to do a special case for when persistence is disabled :) 
(ps. also, can block in JGroups' Flow Control)

Reply:
The former two are fine (data container peek is non blocking and persistence uses a blocking thread if needed). And sadly JGroups blocking is something that can just happen atm, not much we can do. I would propose making this the non blocking executor though.

Reply:
If we have persistence, the non-blocking thread will iterate over all keys and trigger a `persistence.load()` to send the key/value to the remote site. 
Wouldn't it spawn a new blocking thread for each key? if it would, would it be ok?

Reply:
Oh you are manually iterating? And not using keySet or Publisher or anything?

Reply:
It is a ConcurrentHashMap. It contains the keys changes.